### PR TITLE
Make required attributes more flexible

### PR DESCRIPTION
**Pull request by chrisandreae**
Attributes with column defaults are not required by default.
Providing unknown attributes becomes an warning rather than error, in order to better handle new code pre-migration.
Warnings are improved to use AR's logger if present.

### CODE REVIEW

File: lib/persistent_enum.rb [hunk 1]
```diff
@@ -92,20 +92,33 @@ class << self
...
       # populate the constants.
       if model.table_exists?
-        table_attributes = (model.attribute_names - ["id", name_attr.to_s])
-        if required_attributes
-          table_attributes &= required_attributes.map(&:to_s)
+        internal_attributes = ["id", name_attr]
+        table_attributes = (model.attribute_names - internal_attributes)
+
+        # If not otherwise specified, only attributes without defaults are required
+        required_attributes ||= model.column_defaults.map { |attr, default| attr if default.nil? }.compact - internal_attributes
```
Comment (thefloweringash) [inline lib/persistent_enum.rb]: What does this do in the presence of `default NULL` columns? It reads like it should consider them required.

File: lib/persistent_enum.rb [hunk 2]
```diff
@@ -126,11 +139,13 @@ def cache_constants(model, required_members, name_attr: :name, required_attribut
...
-            name = name.to_s
-            attr_names = attrs.keys.map(&:to_s)
+            attr_names = attrs.keys
 
-            if attr_names != table_attributes
-              raise ArgumentError.new("Enum member attribute mismatch: expected #{table_attributes.inspect}, received #{attr_names.inspect}")
+            if (extra_attrs = (attr_names - table_attributes)).present?
+              log_warning("Enum class error: specified attributes #{extra_attrs.inspect} missing from table. Dropping.")
+              attrs = attrs.except(*extra_attrs)
+            elsif (missing_attrs = (required_attributes - attr_names)).present?
```
Comment (thefloweringash) [inline lib/persistent_enum.rb]: This `elsif` reads wrong. When `extra_attrs` are present, surely you still want to check independently that there are no `missing_attrs`?

File: lib/persistent_enum.rb (hunk 1 above)
Comment (chrisandreae) [reply lib/persistent_enum.rb]: It does, which is not desirable. But Rails doesn't seem to collect that information, so it's not easy to detect only the no-default `NOT NULL` case.

File: lib/persistent_enum.rb (hunk 2 above)
Comment (chrisandreae) [reply lib/persistent_enum.rb]: Oops.

**Pull request merged**